feat: Global exception handing 적용

ParkJiwoon opened the pull request:
## Description

Client 가 작업하려면 ErrorHandling 에 대해서 빨리 정의가 필요할 것 같아서 만들었습니다.

기본적인 베이스는 [인스타 때 했던 것](https://bcp0109.tistory.com/303?category=981824)과 동일하고 만들어두신 `ResponseDto` 에 맞춰 추가적으로 작업했습니다.

ParkJiwoon committed:
feat: Global exception handing 적용

- ErrorCode 추가
- ApiException 추가

ParkJiwoon commented on api/src/main/kotlin/org/deepforest/dcinside/auth/AuthService.kt:
```diff
@@ -25,10 +29,13 @@ class AuthService(
-            .also {
-                memberRepository.save(it)
-            }.username
+        val member: Member = signupReqDto.toMember(passwordEncoder)
+
+        if (memberRepository.findByUsername(member.username) != null) {
+            throw ApiException(ErrorCode.CONFLICT_USERNAME)
+        }
+
+        return memberRepository.save(member).username
```
중복 데이터가 있는 경우 Exception 을 던져줘야 하는데 블록을 사용하지 않는게 더 나아서 수정했습니다

ParkJiwoon commented on api/src/main/kotlin/org/deepforest/dcinside/auth/AuthService.kt:
```diff
@@ -38,7 +45,11 @@ class AuthService(
+            authenticationManagerBuilder.getObject().authenticate(authenticationToken)
+        } catch (e: BadCredentialsException) {
+            throw ApiException(ErrorCode.MISMATCH_PASSWORD)
+        }
```
비밀번호 불일치 에러가 나는 경우 우리가 정의한 에러 형식으로 던져주기 위해 적용

DolphaGo replied:
이것도 ControllerAdvice로 옮기면 try~catch를 작성하지 않아도 될 것 같네요.

ParkJiwoon replied:
변경 완료 https://github.com/Project-DC-Inside/dcinside-server/pull/7/commits/b86e1b0915c6501fac41d4a86fec23acaaac286a

ParkJiwoon commented on api/src/test/kotlin/org/deepforest/dcinside/repository/MemberRepositoryTest.kt:
```diff
@@ -19,7 +19,7 @@ class MemberRepositoryTest {
     fun test() {
         memberRepository.save(
-            Member(username = "woody", password = "12345", role = MemberRole.NONE)
+            Member(username = "woody", password = "12345", role = MemberRole.ROLE_NONE)
```
테스트 실패 수정

ParkJiwoon commented on support/src/main/kotlin/org/deepforest/dcinside/dto/Error.kt:
```diff
@@ -1,6 +1,23 @@
-    val code : String,
-    val message: String? = null
-)
+class Error(private val errorCode: ErrorCode) {
+    val code: String
+        get() = errorCode.name
+
+    val message: String
+        get() = errorCode.detail
+}
```
데이터 형식을 조금 변경했습니다.
`ErrorCode` 라는 enum class 를 파라미터로 받으며 외부에 노출할 때는 `code, message` 를 사용합니다.
`get()` 메소드는 Java 의 Getter 라고 생각하시면 됩니다.

자바 스타일로 표현하면 아래와 유사합니다.

```java
public class Error {
    private final ErrorCode errorCode;

    public Error(ErrorCode errorCode) {
        this.errorCode = errorCode;
    }

    public String getCode() {
        return errorCode.name();
    }

    public String getMessage() {
        return errorCode.detail;
    }
}
```

DolphaGo replied:
Setter 닫고 Getter만 열어두신 작업이군요. private 안해도 되나요?

ParkJiwoon replied:
딱히 Setter 를 닫아두려고 한것은 아니고 파라미터로 받는 값과 실제 Getter 로 사용하는 값을 별도로 분리하기 위해서 이렇게 구현했습니다

ParkJiwoon commented on support/src/main/kotlin/org/deepforest/dcinside/dto/ResponseDto.kt:
```diff
@@ -15,8 +15,8 @@ class ResponseDto<R>(
 
-        fun fail(code: String, message: String? = null): ResponseDto<Void> {
-            return ResponseDto(false, error = Error(code, message))
+        fun fail(code: ErrorCode): ResponseDto<Void> {
+            return ResponseDto(false, error = Error(code))
```
파라미터로 ErrorCode 를 받도록 수정했습니다.
별도의 `code`, `message` 를 받지 못하게 되지만 오히려 `ErrorCode` 의 작성을 강제하기 때문에 나쁘지 않다고 생각합니다.

DolphaGo replied:
에러에 대한 정의를 내리도록 강제한다는 점이 나쁘지 않아 보이네요. 좋습니다.

ParkJiwoon reviewed (commented):
이거까지 작업하면 한번 배포 나가도록 하죠

`GlobalExceptionHandler` 의 패키지명을 exception 으로 지었는데 다른 적절한 이름이 있으면 추천해주십쇼.. advice 로..?

DolphaGo commented on api/src/main/kotlin/org/deepforest/dcinside/auth/AuthService.kt:
```diff
@@ -25,10 +29,13 @@ class AuthService(
+            throw ApiException(ErrorCode.CONFLICT_USERNAME)
+        }
+
+        return memberRepository.save(member).username
```
signup시에 username을 반환할 필요는 없어보이네요.(딱히 사용하는 곳이 없어 보여서..)
성공시엔 Void로 success : true 정도만 내어주는 건 어떤가요?

ParkJiwoon replied:
변경 완료 https://github.com/Project-DC-Inside/dcinside-server/pull/7/commits/b86e1b0915c6501fac41d4a86fec23acaaac286a

뭔가 어색하다 싶었는데 Kotlin 에서는 리턴값이 없을 때 `Void` 대신 `Unit` 을 사용합니다.
(참고: https://stackoverflow.com/questions/55953052/kotlin-void-vs-unit-vs-nothing)

DolphaGo commented on api/src/main/kotlin/org/deepforest/dcinside/exception/GlobalExceptionHandler.kt:
```diff
@@ -0,0 +1,20 @@
+package org.deepforest.dcinside.exception
```
Exception 패키지엔 저희가 정의하는 checkedException을 모아두는 용도가 적절한 것 같고 controllerAdvice는 개인적으로 Exception 처리에 대한 설정이기 때문에 `configuration` 패키지가 더 적절한 것 같다고 생각해요.

ParkJiwoon replied:
변경 완료 https://github.com/Project-DC-Inside/dcinside-server/commit/b86e1b0915c6501fac41d4a86fec23acaaac286a

ParkJiwoon committed:
코드리뷰 반영 및 리팩토링

- Void -> Unit 타입으로 변경
- check, checkNotNull 활용
- GlobalExceptionHandler 패키지 이동 및 다른 Exception 추가

ParkJiwoon commented on api/src/main/kotlin/org/deepforest/dcinside/auth/AuthService.kt:
```diff
@@ -24,11 +28,12 @@ class AuthService(
+    fun signup(signupReqDto: SignupReqDto) {
+        val member: Member = signupReqDto.toMember(passwordEncoder)
+
+        check(memberRepository.hasNotUsername(member.username)) { throw ApiException(ErrorCode.CONFLICT_USERNAME) }
```
`check` 라는 메소드를 하나 활용했습니다.

`check(Boolean, () -> Any)` 의 형식으로.. Boolean 값이 false 면 `IllegalStateException` 를 던져주는데..
저희는 그냥 별도의 Exception 을 만들어서 이걸 던져주게 했습니다.

뭔가 어색한것 같으면 그냥 별도로 저희가 하나 만들어서 사용하면 될거같네요

DolphaGo replied:
코틀린 내장이에요?

ParkJiwoon replied:
네 check, checkNotNull, require, requireNotNull 이렇게 지원합니다

DolphaGo replied:
```kotlin
check (Predicate){
  // false 시에 동작하는 코드
}
```
인가 보네요.


inline 말고 newLine 하나 넣어주죠

ParkJiwoon replied:
원래 사용법은 이런 스타일이긴 합니다 

```kt
check (Boolean) { "중복된 ID 가 존재합니다" }
```

DolphaGo replied:
껄껄 그렇군요. 그럼 패스~! ㅎㅎ

ParkJiwoon replied:
우리는 저대로 사용 안하니까 변경했습니다 https://github.com/Project-DC-Inside/dcinside-server/pull/7/commits/db32fc4363689ac31efa574df6168b07eb5a6826

DolphaGo reviewed (approved):
👍🏻

ParkJiwoon committed:
코드리뷰 반영

- UsernmaeNotFoundException 을 ExceptionHandler 에서 못잡아서 롤백
- check 메소드에 new line 추가
